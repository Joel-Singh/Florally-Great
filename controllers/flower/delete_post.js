const asyncHandler = require("express-async-handler");
const { default: Flower } = require("../../models/flower.ts");
const he = require("he");
module.exports = asyncHandler(async (req, res, next) => {
  const { flowerId } = req.body;
  await Flower.findByIdAndDelete(flowerId);
  const name = he.decode(req.params.name);

  res.render("message", {
    title: `${name || "Flower"} deleted!`,
    message: `${name || "Flower"} deleted successfully`,
  });
});
