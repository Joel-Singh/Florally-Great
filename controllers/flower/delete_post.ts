import asyncHandler from "express-async-handler";
import Flower from "../../models/flower.ts";

import { decode } from "he";

export default asyncHandler(async (req, res, next) => {
  //@ts-ignore
  const { flowerId } = req.body;
  await Flower.findByIdAndDelete(flowerId);
  //@ts-ignore
  const name = decode(req.params.name);

  //@ts-ignore
  res.render("message", {
    title: `${name || "Flower"} deleted!`,
    message: `${name || "Flower"} deleted successfully`,
  });
});
