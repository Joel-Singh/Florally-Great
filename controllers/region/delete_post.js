const asyncHandler = require("express-async-handler");
const path = require("path");
const Region = require(path.join(appRoot, "models", "region.js"));
const Flower = require(path.join(appRoot, "models", "flower.js"));
const renderDeleteRegion = require("./rendersWithDefaultLocals/renderDeleteRegion.js");

module.exports = asyncHandler(async (req, res, next) => {
  if (typeof req.body.region === 'undefined') {
    renderDeleteRegion(res, {
      errors: [{
        msg: "Please select a region"
      }]
    });
    return;
  }

  const regionId = req.body.region;

  if (!(await regionHasFlower(regionId))) {
    await Region.findByIdAndDelete(regionId);
    res.redirect("regions/delete");
  } else {
    const flowersOfRegion = await Flower.find({ region: regionId }).exec();
    const errors = [
      {
        msg: createFlowerRegionError(flowersOfRegion),
      },
    ];

    renderDeleteRegion(res, {
      errors,
    });
  }
});

function createFlowerRegionError(flowersOfRegion) {
  const flowersAsAnchorsInListItem = flowersOfRegion.map((flower) => {
    const { url, name } = flower;
    return `<li><a href="${url}">${name}</a></li>`;
  });

  const allListItemsInSingleString = flowersAsAnchorsInListItem.reduce(
    (previous, current) => {
      return previous + current;
    }
  );

  const unorderedListOfFlowers = "<ul>" + allListItemsInSingleString + "</ul>";

  const header = `<h2>Region has flowers, delete them first: </h2>`;

  return header + unorderedListOfFlowers;
}

async function regionHasFlower(regionId) {
  const flower = await Flower.findOne({ region: regionId }).exec();
  if (flower === null) return false;
  else return true;
}
