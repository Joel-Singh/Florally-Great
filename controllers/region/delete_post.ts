import asyncHandler from "express-async-handler";
import renderDeleteRegion from "./rendersWithDefaultLocals/renderDeleteRegion.ts";

import Flower from "../../models/flower";
import Region from "../../models/region";
import { RequestHandler } from "express";
import {
  RegionDeleteFormData,
  RequestWithRegionDeleteFormData,
} from "../../views/regions/regionDeleteFormInterfaces.ts";
import { Response } from "express-serve-static-core";

const delete_post: RequestHandler = async (
  req: RequestWithRegionDeleteFormData,
  res,
  next
) => {
  const { regionId, fromRegionDetailPage } = req.body;

  if (typeof regionId === "undefined") {
    await renderDeleteRegion(res, {
      errors: [
        {
          msg: "Please select a region",
        },
      ],
    });
    return;
  }

  const region = await Region.findById(regionId);
  if (region === null) {
    throw new Error("Couldn't find region of requested deleted region");
  }
  const regionName = region.name;

  if (await regionHasFlower(regionId)) {
    await regionDeleteFailureRender(
      regionId,
      res,
      fromRegionDetailPage,
      regionName
    );
  } else {
    await Region.findByIdAndDelete(regionId);
    await regionDeleteSuccessRender(res, fromRegionDetailPage, regionName);
  }
};

export default asyncHandler(delete_post as any);

async function regionDeleteFailureRender(
  regionId: string,
  res: Response,
  fromRegionDetailPage: RegionDeleteFormData["fromRegionDetailPage"],
  regionName: string
) {
  if (fromRegionDetailPage) {
    res.render("message", {
      title: "Failure!",
      message: `${regionName} not deleted`,
    });
    return;
  } else {
    await renderDeleteRegionWithFlowerError(regionId, res);
    return;
  }
}

async function regionDeleteSuccessRender(
  res,
  fromRegionDetailPage: RegionDeleteFormData["fromRegionDetailPage"],
  regionName: string
) {
  if (fromRegionDetailPage) {
    res.render("message", {
      title: "Success!",
      message: `${regionName} successfully deleted`,
    });
  } else {
    res.redirect("/regions/delete");
    return;
  }
}

async function regionHasFlower(regionId) {
  const flower = await Flower.findOne({ region: regionId }).exec();
  if (flower === null) return false;
  else return true;
}

async function renderDeleteRegionWithFlowerError(regionId, res) {
  const flowersOfRegion = await Flower.find({ region: regionId }).exec();
  const errors = [
    {
      msg: createFlowerRegionError(flowersOfRegion),
    },
  ];

  await renderDeleteRegion(res, {
    errors,
  });

  function createFlowerRegionError(flowersOfRegion) {
    const flowersAsAnchorsInListItem = flowersOfRegion.map((flower) => {
      const { url, name } = flower;
      return `<li><a href="${url}">${name}</a></li>`;
    });

    const allListItemsInSingleString = flowersAsAnchorsInListItem.reduce(
      (previous, current) => {
        return previous + current;
      }
    );

    const unorderedListOfFlowers =
      "<ul>" + allListItemsInSingleString + "</ul>";

    const header = `<h2>Region has flowers, delete them first: </h2>`;

    return header + unorderedListOfFlowers;
  }
}
