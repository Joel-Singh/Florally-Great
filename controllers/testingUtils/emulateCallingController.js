module.exports = async function (controller, reqProperties, resProperties) {
  const { fakeReq, fakeRes } = getFakeMiddleware(reqProperties, resProperties);

  await controller(fakeReq, fakeRes);

  return {
    fakeReq,
    fakeRes,
    getRenderInformation,
  };

  function getFakeMiddleware(options) {
    const fakeReq = {
      ...reqProperties,
    };

    const fakeRes = {
      render: jest.fn(),
      redirect: jest.fn(),
      ...resProperties,
    };

    return { fakeReq, fakeRes };
  }
};

function getRenderInformation(fakeRes) {
  const renderCalls = fakeRes.render.mock.calls;
  if (renderCalls.length === 0) throw new Error("Render hasn't been called!");

  const renderCall = renderCalls[0];
  return {
    view: renderCall[0],
    locals: renderCall[1],
  };
}
