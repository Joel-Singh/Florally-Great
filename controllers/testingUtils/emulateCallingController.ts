import { Request, Response } from "express";

type nestedFunctionArray = Array<Function | nestedFunctionArray>;
export default async function emulateCallingController(
  controller: nestedFunctionArray | Function,
  reqProperties: Partial<Request> = {},
  resProperties: Partial<Response> = {},
) {
  const { fakeReq, fakeRes } = getFakeMiddleware(reqProperties, resProperties);
  const mockNext = jest.fn();

  if (!Array.isArray(controller)) {
    await controller(fakeReq, fakeRes, mockNext);
  } else {
    // @ts-ignore
    controller = controller.flat(Infinity) as Array<Function>;
    await runMiddlewareArray(
      controller as Array<Function>,
      fakeReq,
      fakeRes,
      mockNext,
    );
  }

  return {
    fakeReq,
    fakeRes,
    mockNext,
    getRenderInformation: createMockInfoGetter(
      fakeRes.render,
      "view",
      "locals",
    ),
    getRedirectInformation: createMockInfoGetter(
      fakeRes.redirect,
      "redirectPage",
    ),
    getMockNextInformation: createMockInfoGetter(mockNext, "errorOrEmpty"),
  };
}

function getFakeMiddleware(
  reqProperties: Partial<Request>,
  resProperties: Partial<Response>,
) {
  const fakeReq = {
    ...reqProperties,
  };

  const fakeRes = {
    ...resProperties,
    render: jest.fn(),
    redirect: jest.fn(),
    send: jest.fn(),
  };

  return { fakeReq, fakeRes };
}

function createMockInfoGetter<T extends Record<string, any>>(
  mockFunction: jest.Mock,
  ...namesInArgumentOrder: Array<keyof T>
) {
  return function () {
    const mockCalls = mockFunction.mock.calls;
    if (mockCalls.length === 0)
      throw new Error("Mock function hasn't been called!");

    const mockCall = mockCalls[0];
    const mockInfo: Partial<T> = {};

    namesInArgumentOrder.forEach((name, index) => {
      mockInfo[name] = mockCall[index];
    });

    return mockInfo as T;
  };
}

async function runMiddlewareArray(
  middlewares: Array<Function>,
  req: Partial<Request>,
  res: Partial<Response>,
  next: ReturnType<typeof jest.fn>,
) {
  for (const middleware of middlewares) {
    let previousCallLength = next.mock.calls.length;
    await middleware(req, res, next);
    let currentCallLength = next.mock.calls.length;
    let nextWasCalled = currentCallLength > previousCallLength;
    if (!nextWasCalled) break;
  }
}
