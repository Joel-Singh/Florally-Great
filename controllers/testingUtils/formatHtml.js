const prettier = require("prettier");

module.exports = function (html) {
  return prettier.format(html, {
    parser: "html",
  });
};
