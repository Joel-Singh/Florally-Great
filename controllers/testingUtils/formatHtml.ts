import { format } from "prettier";

export default function (html) {
  return format(html, {
    parser: "html",
  });
}
