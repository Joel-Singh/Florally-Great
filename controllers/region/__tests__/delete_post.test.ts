import formatHtml from "../../testingUtils/formatHtml.js";
import { default as emulateCallingController } from "../../testingUtils/emulateCallingController";

import Flower from "../../../models/flower";
import Region from "../../../models/region";

import delete_post from "../delete_post";
import {
  RegionDeleteFormData,
  RequestWithRegionDeleteFormData,
} from "../../../views/regions/regionDeleteFormInterfaces.js";

test("If no region is selected, error message is returned", async () => {
  const { getRenderInformation } = await emulateCallingController(delete_post, {
    body: { regionId: undefined },
  });

  const { view, locals } = getRenderInformation();

  expect(view).toMatchInlineSnapshot(`"regions/delete_region"`);
  expect(locals.errors).toMatchInlineSnapshot(`
    [
      {
        "msg": "Please select a region",
      },
    ]
  `);
});

describe("On region without a flower", () => {
  test("Region is deleted", async () => {
    const id = await saveRegionToBeDeleted();

    await emulateCallingController(delete_post, {
      body: { regionId: id },
    });

    const foundRegion = await Region.findById(id).exec();
    expect(foundRegion).toBeNull();
  });

  test("Renders form again", async () => {
    const id = await saveRegionToBeDeleted();

    const { getRedirectInformation } = await emulateCallingController(
      delete_post,
      {
        body: { regionId: id },
      }
    );

    const { redirectPage } = getRedirectInformation();
    expect(redirectPage).toMatchInlineSnapshot(`"/regions/delete"`);
  });
});

describe("On region with a flower", () => {
  test("Region isn't deleted when it has flower", async () => {
    const id = await saveRegionWithFlower();

    await emulateCallingController(delete_post, {
      body: { regionId: id },
    });

    const foundRegion = await Region.findById(id).exec();
    expect(foundRegion).not.toBeNull();
  });

  test("Renders Error, hyperlinking to flower", async () => {
    const id = await saveRegionWithFlower();

    const { getRenderInformation } = await emulateCallingController(
      delete_post,
      {
        body: { regionId: id },
      }
    );

    const { locals } = getRenderInformation();

    const errorMsg = locals.errors[0].msg;
    expect(formatHtml(errorMsg)).toMatchSnapshot();
  });

  test("Form after error rerender still renders with regions", async () => {
    const id = await saveRegionWithFlower();

    const regionShouldShowUp = new Region({
      name: "I should be in the snapshot!",
      description: "Hello! you should see this!",
      _id: "647cbdb6ec4dd230927acfdf",
    });

    await regionShouldShowUp.save();

    const { getRenderInformation } = await emulateCallingController(
      delete_post,
      {
        body: { regionId: id },
      }
    );

    const { locals } = getRenderInformation();

    expect(locals.all_regions).toMatchSnapshot();
  });
});

describe("On region with multiple flowers", () => {
  test("Renders error, hyperlinking to multiple flowers", async () => {
    const id = await saveRegionToBeDeleted();

    await addFlowerToRegion("Flower 1", id);
    await addFlowerToRegion("Flower 2", id);
    await addFlowerToRegion("Flower 3", id);

    const { getRenderInformation } = await emulateCallingController(
      delete_post,
      {
        body: { regionId: id },
      }
    );

    const { locals } = getRenderInformation();

    const errorMsg = locals.errors[0].msg;
    expect(formatHtml(errorMsg)).toMatchSnapshot();

    async function addFlowerToRegion(name, id) {
      const flowerInRegion = new Flower({
        name,
        description: "desc",
        price: "32",
        numberInStock: "1",
        region: id,
      });
      await flowerInRegion.save();
    }
  });
});

describe("From deleting through region detail page", () => {
  test("If region doesn't have flowers, deletes region and redirects with success message", async () => {
    const id = await saveRegionToBeDeleted();
    const body: RegionDeleteFormData = {
      regionId: id.toString(),
      fromRegionDetailPage: "true",
    };

    const { getRenderInformation } = await emulateCallingController(
      delete_post,
      {
        body,
      }
    );

    const { view, locals } = getRenderInformation();

    const foundRegion = await Region.findById(id).exec();
    expect(foundRegion).toBeNull();

    expect(view).toMatchInlineSnapshot(`"message"`);
    expect(locals).toMatchInlineSnapshot(`
      {
        "message": "Region successfully deleted",
        "title": "success!",
      }
    `);
  });

  test("If region does have flowers, don't delete region and redirect with error", async () => {
    const id = await saveRegionWithFlower();
    const { getRenderInformation } = await emulateCallingController(
      delete_post,
      {
        body: {
          regionId: id,
          fromRegionDetailPage: true,
        },
      }
    );

    const { view, locals } = getRenderInformation();

    const foundRegion = await Region.findById(id).exec();
    expect(foundRegion).not.toBeNull();

    expect(view).toMatchInlineSnapshot(`"message"`);
    expect(locals).toMatchInlineSnapshot(`
      {
        "message": "Region not deleted",
        "title": "failure!",
      }
    `);
  });
});

async function saveRegionWithFlower() {
  const id = await saveRegionToBeDeleted();

  const flowerInRegion = new Flower({
    name: "Flower",
    description: "desc",
    price: "32",
    numberInStock: "1",
    region: id,
  });

  await flowerInRegion.save();

  return id;
}

async function saveRegionToBeDeleted() {
  const regionToBeDeleted = new Region({
    name: "Region Name",
    description: "I'm gonna be deleted!",
    _id: "647cbde08678de7aaf63d2ec",
  });

  const regionToBeDeletedDoc = await regionToBeDeleted.save();
  const id = regionToBeDeletedDoc._id;

  return id;
}
