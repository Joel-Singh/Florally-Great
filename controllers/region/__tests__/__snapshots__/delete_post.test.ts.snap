// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`On region with a flower Form after error rerender still renders with regions 1`] = `
[
  {
    "_id": "647cbde08678de7aaf63d2ec",
    "name": "Region Name",
  },
  {
    "_id": "647cbdb6ec4dd230927acfdf",
    "name": "I should be in the snapshot!",
  },
]
`;

exports[`On region with a flower Renders Error, hyperlinking to flower 1`] = `
"<h2>Region has flowers, delete them first:</h2>
<ul>
  <li><a href="/flowers/Flower">Flower</a></li>
</ul>
"
`;

exports[`On region with multiple flowers Renders error, hyperlinking to multiple flowers 1`] = `
"<h2>Region has flowers, delete them first:</h2>
<ul>
  <li><a href="/flowers/Flower 1">Flower 1</a></li>
  <li><a href="/flowers/Flower 2">Flower 2</a></li>
  <li><a href="/flowers/Flower 3">Flower 3</a></li>
</ul>
"
`;
